[Bug] `docker compose build` check fails but only on GitHub

# Expected Behavior

The `docker compose build` should build the container with no problems

# Actual Behavior

The build fails because `psychopg2` doesn't seem to have a binary so it tries to build from scratch, which fails.

# Steps to Reproduce the Problem

1. Run `docker compose build` locally on your machine, which would probably work.
1. Upon trying to do a PR, it will fail on GitHub. (You can make a draft PR if you want, or look at GitHub Actions' documentation on how to recreate their environment.)

# Additional Info

The two options I'm thinking to fix it (but might be right or wrong):
1. Attempt to upgrade `psychopg2` or `psychopg2-binary` if they're available
2. Convert `Dockerfile` to a two stage build. Attempt to add in the build tools during the first step of the build, then the second stage can be just taking the code post-build and slimming down the container.

<details>
<summary>Error log</summary>

```
  docker compose build
  shell: /usr/bin/bash -e {0}
time="2025-01-28T16:38:27Z" level=warning msg="/home/runner/work/codethesaur.us/codethesaur.us/docker-compose.yml: `version` is obsolete"
#0 building with "default" instance using docker driver

#1 [app internal] load build definition from Dockerfile
#1 transferring dockerfile: 331B done
#1 DONE 0.0s

#2 [app auth] library/python:pull token for registry-1.docker.io
#2 DONE 0.0s

#3 [app internal] load metadata for docker.io/library/python:3-slim
#3 DONE 0.6s

#4 [app internal] load .dockerignore
#4 transferring context: 114B done
#4 DONE 0.0s

#5 [app internal] load build context
#5 transferring context: 2.27MB 0.0s done
#5 DONE 0.1s

#6 [app 1/5] FROM docker.io/library/python:3-slim@sha256:026dd417a88d0be8ed5542a05cff5979d17625151be8a1e25a994f85c87962a5
#6 resolve docker.io/library/python:3-slim@sha256:026dd417a88d0be8ed5542a05cff5979d17625151be8a1e25a994f85c87962a5 done
#6 sha256:311aef7e53025f72dd19d6eff1f5c8a899e8fe5d1ccc1d56bae4a6ec514325cf 5.17kB / 5.17kB done
#6 sha256:af302e5c37e9dc1dbe2eadc8f5059d82a914066b541b0d1a6daa91d0cc55057d 5.24MB / 28.21MB 0.1s
#6 sha256:5ac68359aa43fab9969e318e2f48a80e3da0b46266608fb7d428ed35825c600e 0B / 3.51MB 0.1s
#6 sha256:7c98ec0dbd9ce94ea3cd695c089efeda24fffa0d545b5bb9b4003250bad195e5 0B / 12.58MB 0.1s
#6 sha256:026dd417a88d0be8ed5542a05cff5979d17625151be8a1e25a994f85c87962a5 9.12kB / 9.12kB done
#6 sha256:5e73d1759f8efcf72d00f78ef9f8826959c88e7edcd22b558e3006a5b325f03b 1.75kB / 1.75kB done
#6 sha256:af302e5c37e9dc1dbe2eadc8f5059d82a914066b541b0d1a6daa91d0cc55057d 28.21MB / 28.21MB 0.3s done
#6 sha256:5ac68359aa43fab9969e318e2f48a80e3da0b46266608fb7d428ed35825c600e 3.51MB / 3.51MB 0.2s done
#6 sha256:7c98ec0dbd9ce94ea3cd695c089efeda24fffa0d545b5bb9b4003250bad195e5 12.58MB / 12.58MB 0.3s done
#6 sha256:4b4f99cb7dcbff680f31a8fbe699fcb5d988a526d54b8e7c6e927f4ef51bc920 251B / 251B 0.3s done
#6 extracting sha256:af302e5c37e9dc1dbe2eadc8f5059d82a914066b541b0d1a6daa91d0cc55057d 0.1s
#6 extracting sha256:af302e5c37e9dc1dbe2eadc8f5059d82a914066b541b0d1a6daa91d0cc55057d 1.0s done
#6 extracting sha256:5ac68359aa43fab9969e318e2f48a80e3da0b46266608fb7d428ed35825c600e
#6 extracting sha256:5ac68359aa43fab9969e318e2f48a80e3da0b46266608fb7d428ed35825c600e 0.1s done
#6 extracting sha256:7c98ec0dbd9ce94ea3cd695c089efeda24fffa0d545b5bb9b4003250bad195e5
#6 extracting sha256:7c98ec0dbd9ce94ea3cd695c089efeda24fffa0d545b5bb9b4003250bad195e5 0.5s done
#6 extracting sha256:4b4f99cb7dcbff680f31a8fbe699fcb5d988a526d54b8e7c6e927f4ef51bc920
#6 extracting sha256:4b4f99cb7dcbff680f31a8fbe699fcb5d988a526d54b8e7c6e927f4ef51bc920 done
#6 DONE 2.2s

#7 [app 2/5] WORKDIR /code
#7 DONE 0.0s

#8 [app 3/5] COPY requirements.txt /code/
#8 DONE 0.0s

#9 [app 4/5] RUN pip install --no-cache-dir -r requirements.txt
#9 1.543 Collecting asgiref==3.7.2 (from -r requirements.txt (line 1))
#9 1.629   Downloading asgiref-3.7.2-py3-none-any.whl.metadata (9.2 kB)
#9 1.693 Collecting astroid==3.0.1 (from -r requirements.txt (line 2))
#9 1.713   Downloading astroid-3.0.1-py3-none-any.whl.metadata (4.5 kB)
#9 1.732 Collecting attrs==23.1.0 (from -r requirements.txt (line 3))
#9 1.752   Downloading attrs-23.1.0-py3-none-any.whl.metadata (11 kB)
#9 1.789 Collecting bleach==6.1.0 (from -r requirements.txt (line 4))
#9 1.809   Downloading bleach-6.1.0-py3-none-any.whl.metadata (30 kB)
#9 1.843 Collecting colorama==0.4.6 (from -r requirements.txt (line 5))
#9 1.863   Downloading colorama-0.4.6-py2.py3-none-any.whl.metadata (17 kB)
#9 1.879 Collecting dill==0.3.7 (from -r requirements.txt (line 6))
#9 1.899   Downloading dill-0.3.7-py3-none-any.whl.metadata (9.9 kB)
#9 1.910 Collecting dj-database-url==2.1.0 (from -r requirements.txt (line 7))
#9 1.930   Downloading dj_database_url-2.1.0-py3-none-any.whl.metadata (11 kB)
#9 2.028 Collecting Django==4.2.17 (from -r requirements.txt (line 8))
#9 2.048   Downloading Django-4.2.17-py3-none-any.whl.metadata (4.1 kB)
#9 2.058 Collecting django-markdownify==0.9.3 (from -r requirements.txt (line 9))
#9 2.079   Downloading django_markdownify-0.9.3-py3-none-any.whl.metadata (2.4 kB)
#9 2.109 Collecting django-on-heroku==1.1.2 (from -r requirements.txt (line 10))
#9 2.129   Downloading django_on_heroku-1.1.2-py2.py3-none-any.whl.metadata (3.6 kB)
#9 2.149 Collecting gunicorn==22.0.0 (from -r requirements.txt (line 11))
#9 2.169   Downloading gunicorn-22.0.0-py3-none-any.whl.metadata (4.4 kB)
#9 2.231 Collecting importlib-metadata==6.8.0 (from -r requirements.txt (line 12))
#9 2.251   Downloading importlib_metadata-6.8.0-py3-none-any.whl.metadata (5.1 kB)
#9 2.299 Collecting isort==5.12.0 (from -r requirements.txt (line 13))
#9 2.319   Downloading isort-5.12.0-py3-none-any.whl.metadata (12 kB)
#9 2.330 Collecting jsonmerge==1.9.2 (from -r requirements.txt (line 14))
#9 2.351   Downloading jsonmerge-1.9.2-py3-none-any.whl.metadata (21 kB)
#9 2.386 Collecting jsonschema==4.19.2 (from -r requirements.txt (line 15))
#9 2.406   Downloading jsonschema-4.19.2-py3-none-any.whl.metadata (7.9 kB)
#9 2.423 Collecting jsonschema-specifications==2023.7.1 (from -r requirements.txt (line 16))
#9 2.442   Downloading jsonschema_specifications-2023.7.1-py3-none-any.whl.metadata (2.8 kB)
#9 2.482 Collecting lazy-object-proxy==1.9.0 (from -r requirements.txt (line 17))
#9 2.502   Downloading lazy-object-proxy-1.9.0.tar.gz (42 kB)
#9 2.528   Installing build dependencies: started
#9 4.555   Installing build dependencies: finished with status 'done'
#9 4.557   Getting requirements to build wheel: started
#9 4.735   Getting requirements to build wheel: finished with status 'done'
#9 4.737   Preparing metadata (pyproject.toml): started
#9 4.947   Preparing metadata (pyproject.toml): finished with status 'done'
#9 4.969 Collecting Markdown==3.5 (from -r requirements.txt (line 18))
#9 4.990   Downloading Markdown-3.5-py3-none-any.whl.metadata (7.1 kB)
#9 5.000 Collecting mccabe==0.7.0 (from -r requirements.txt (line 19))
#9 5.020   Downloading mccabe-0.7.0-py2.py3-none-any.whl.metadata (5.0 kB)
#9 5.045 Collecting packaging==23.2 (from -r requirements.txt (line 20))
#9 5.065   Downloading packaging-23.2-py3-none-any.whl.metadata (3.2 kB)
#9 5.089 Collecting platformdirs==3.11.0 (from -r requirements.txt (line 21))
#9 5.109   Downloading platformdirs-3.11.0-py3-none-any.whl.metadata (11 kB)
#9 5.314 Collecting protobuf==4.24.4 (from -r requirements.txt (line 22))
#9 5.334   Downloading protobuf-4.24.4-cp37-abi3-manylinux2014_x86_64.whl.metadata (540 bytes)
#9 5.408 Collecting psycopg2-binary==2.9.9 (from -r requirements.txt (line 23))
#9 5.429   Downloading psycopg2-binary-2.9.9.tar.gz (384 kB)
#9 5.529   Installing build dependencies: started
#9 6.972   Installing build dependencies: finished with status 'done'
#9 6.973   Getting requirements to build wheel: started
#9 7.151   Getting requirements to build wheel: finished with status 'error'
#9 7.156   error: subprocess-exited-with-error
#9 7.156   
#9 7.156   × Getting requirements to build wheel did not run successfully.
#9 7.156   │ exit code: 1
#9 7.156   ╰─> [21 lines of output]
#9 7.156       running egg_info
#9 7.156       writing psycopg2_binary.egg-info/PKG-INFO
#9 7.156       writing dependency_links to psycopg2_binary.egg-info/dependency_links.txt
#9 7.156       writing top-level names to psycopg2_binary.egg-info/top_level.txt
#9 7.156       
#9 7.156       Error: pg_config executable not found.
#9 7.156       
#9 7.156       pg_config is required to build psycopg2 from source.  Please add the directory
#9 7.156       containing pg_config to the $PATH or specify the full executable path with the
#9 7.156       option:
#9 7.156       
#9 7.156           python setup.py build_ext --pg-config /path/to/pg_config build ...
#9 7.156       
#9 7.156       or with the pg_config option in 'setup.cfg'.
#9 7.156       
#9 7.156       If you prefer to avoid building psycopg2 from source, please install the PyPI
#9 7.156       'psycopg2-binary' package instead.
#9 7.156       
#9 7.156       For further information please check the 'doc/src/install.rst' file (also at
#9 7.156       <https://www.psycopg.org/docs/install.html>).
#9 7.156       
#9 7.156       [end of output]
#9 7.156   
#9 7.156   note: This error originates from a subprocess, and is likely not a problem with pip.
#9 7.289 
#9 7.289 [notice] A new release of pip is available: 24.3.1 -> 25.0
#9 7.289 [notice] To update, run: pip install --upgrade pip
#9 7.291 error: subprocess-exited-with-error
#9 7.291 
#9 7.291 × Getting requirements to build wheel did not run successfully.
#9 7.291 │ exit code: 1
#9 7.291 ╰─> See above for output.
#9 7.291 
#9 7.291 note: This error originates from a subprocess, and is likely not a problem with pip.
#9 ERROR: process "/bin/sh -c pip install --no-cache-dir -r requirements.txt" did not complete successfully: exit code: 1
------
 > [app 4/5] RUN pip install --no-cache-dir -r requirements.txt:
7.289 
Notice: 7.289 [notice] A new release of pip is available: 24.3.1 -> 25.0
Notice: 7.289 [notice] To update, run: pip install --upgrade pip
7.291 error: subprocess-exited-with-error
7.291 
7.291 × Getting requirements to build wheel did not run successfully.
7.291 │ exit code: 1
7.291 ╰─> See above for output.
7.291 
7.291 note: This error originates from a subprocess, and is likely not a problem with pip.
------
failed to solve: process "/bin/sh -c pip install --no-cache-dir -r requirements.txt" did not complete successfully: exit code: 1
Error: Process completed with exit code 17.
```

</details>

--- a/requirements.txt
+++ b/requirements.txt
@@ -20,13 +20,13 @@ mccabe==0.7.0
 packaging==23.2
 platformdirs==3.11.0
 protobuf==4.24.4
-psycopg2-binary==2.9.9
+psycopg2-binary==2.9.10
 Pygments==2.16.1
 pylint==3.0.2
 pyparsing==3.1.1
 pytz==2023.3.post1
 referencing==0.30.2
-rpds-py==0.10.6
+rpds-py==0.22.3
 six==1.16.0
 sqlparse==0.5.0
 tinycss2==1.2.1

